services/brig/test/integration/API/OAuth.hs

```diff
@@ -0,0 +1,780 @@
+          test m "wrong signature - fail" $ testRefreshTokenWrongSignature o b,
+          test m "no token id - fail" $ testRefreshTokenNoTokenId o b,
+          test m "non-existing id - fail" $ testRefreshTokenNonExistingId o b,
+          test m "wrong client id - fail" $ testRefreshTokenWrongClientId b,
+          test m "wrong client secret - fail" $ testRefreshTokenWrongClientSecret b,
+          test m "wrong grant type - fail" $ testRefreshTokenWrongGrantType b,
+          test m "expired token - fail" $ testRefreshTokenExpiredToken o b,
+          test m "revoked token - fail" $ testRefreshTokenRevokedToken b
+        ],
+      testGroup "oauth applications" $
+        [ test m "list applications with account access" $ testListApplicationsWithAccountAccess b,
+          test m "revoke application account access" $ testRevokeApplicationAccountAccess b
+        ]
+    ]
+
+testRegisterNewOAuthClient :: Brig -> Http ()
+testRegisterNewOAuthClient brig = do
+  let newOAuthClient@(NewOAuthClient expectedAppName expectedUrl) = newOAuthClientRequestBody "E Corp" "https://example.com"
+  cid <- occClientId <$> registerNewOAuthClient brig newOAuthClient
+  uid <- randomId
+  oauthClientInfo <- getOAuthClientInfo brig uid cid
+  liftIO $ do
+    expectedAppName @?= ocName oauthClientInfo
+    expectedUrl @?= ocRedirectUrl oauthClientInfo
+
+testCreateOAuthCodeSuccess :: Brig -> Http ()
+testCreateOAuthCodeSuccess brig = do
+  let newOAuthClient@(NewOAuthClient _ redirectUrl) = newOAuthClientRequestBody "E Corp" "https://example.com"
+  cid <- occClientId <$> registerNewOAuthClient brig newOAuthClient
+  uid <- randomId
+  let scope = OAuthScopes $ Set.fromList [ConversationCreate, ConversationCodeCreate]
+  let state = "foobar"
+  createOAuthCode brig uid (NewOAuthAuthCode cid scope OAuthResponseTypeCode redirectUrl state) !!! do
+    const 302 === statusCode
+    const (Just $ unRedirectUrl redirectUrl ^. pathL) === (fmap getPath . getLocation)
+    const (Just $ ["code", "state"]) === (fmap (fmap fst . getQueryParams) . getLocation)
+    const (Just $ cs state) === (getLocation >=> getQueryParamValue "state")
+
+testCreateOAuthCodeRedirectUrlMismatch :: Brig -> Http ()
+testCreateOAuthCodeRedirectUrlMismatch brig = do
+  let newOAuthClient@(NewOAuthClient _ redirectUrl) = newOAuthClientRequestBody "E Corp" "https://example.com"
+  cid <- occClientId <$> registerNewOAuthClient brig newOAuthClient
+  uid <- randomId
+  let state = "foobar"
+  let differentUrl = mkUrl "https://wire.com"
+  createOAuthCode brig uid (NewOAuthAuthCode cid (OAuthScopes Set.empty) OAuthResponseTypeCode differentUrl state) !!! do
+    const 400 === statusCode
+    const (Just $ "access_denied") === (getLocation >=> getQueryParamValue "error")
+    const (Just $ cs state) === (getLocation >=> getQueryParamValue "state")
+    const (Just $ unRedirectUrl redirectUrl ^. pathL) === (fmap getPath . getLocation)
```

Comment:
should this be Nothing as well?  is it bad that we give away information about the client that appears to be under attack here?  (same in a few places below.)

Reply:
No, it's correct according to the specs. We do not reveal any additional information, since the returned redirect URL is the same as was passed in the request by the client. The redirect is, in fact, the mechanism to communicate back to the client. That's why errors are added as query parameters.

Reply:
I don't understand.  `redirectUrl` is part of the oauth client's identity, communicated between oauth client and brig.  the `createOAuthCode` request can be sent by an attacker who knows the client id, but not the redirect url, and we want to avoid leaking the redirect url, given that the sender clearly doesn't know it.

I don't think this is a big deal, just that it's not necessary to reveal that information here, and (at least theoretically) harmful.

Reply:
Yes, the redirect URL is part of the client's identity. And it is additionally provided by the client with the authorization request. The redirect URL that is passed has to match the one that was registered with the client. And the URL that has been passed is returned, not the one registered with the client.

Reply:
[from the specs](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.2.1)

> If the request fails due to a missing, invalid, or mismatching
   redirection URI, or if the client identifier is missing or invalid,
   the authorization server SHOULD inform the resource owner of the
   error and MUST NOT automatically redirect the user-agent to the
   invalid redirection URI.
   
so you are right, we must not redirect but return an error.